ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+### Return values and errors
+
+An activity can communicate its result to the workflow by returning a value or raising an error. The
+return values will get serialized as a `Payload` object similar to activity input arguments.
+Similarly, an error will get serialized and re-raised from within the workflow (if an activity
+called "synchronously").
+
+The main idea here is to make it feel (from the workflow's perspective) like a local invocation of
+an activity without any additional constructs in the way.
+
+```ruby
+class MakePaymentActivity < Temporal::Activity
+  PaymentMethodNotFound = Class.new(Temporal::Error::ActivityFailure)
+  PaymentFailed = Class.new(Temporal::Error::ActivityFailure)
+
+  def execute(payment_method_id)
+    payment_method = PaymentMethod.find(payment_method_id)
+    raise UserNotFound, "No payment_method with id #{payment_method_id}" unless payment_method
+
+    result = PaymentService.execute(user.payment_method)
+    raise PaymentFailed, "Unable to process payment: #{result.error}" if result.failure?
+
+    result.payment_id
+  end
+end
+```
+
+Notes:
+
+- Errors defined within the activity class are namespaced as `MyActivity::PaymentFailed`,
+  which is a very useful property
+  - This will however not work if a workflow does not have access to the activity definition
+  - In this case it will get raised as a generic `Temporal::Error::ActivityFailure` with an original
+    error's class name available via `#original_error`
+- Subclassing `Temporal::Error::ActivityFailure` allows us to inject extra information into the
+  error (such as activity name, id, etc)
+- All errors will also have their backtrace serialized and relayed to the caller
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
+
+The main difference is that a DataConverter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the later is for
+modifying a binary that is transmitted over the wire.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7933')
+
+config.add_data_converter(MyProtobufConverter.new)
+config.add_data_converter(MyThriftConverter) # can be a class/module based on duck-typing
+
+config.add_codec(EncryptionCodec.new(KEY_ID))
+config.add_codec(CompressionCodec)
+```
+
+A data converter is expected to have this interface:
+
+```ruby
+class Temporal::DataConverter::Base
+  # Returns mime-type of the payload
+  def encoding() -> String
```

Comment:
Not all uses of data converters are encoding based. That's just convention for multi-payload-converter data converters. We can kinda try to force it here, but really, they could just choose to ignore the idea of encoding altogether.

Reply:
This was my attempt to make it visible when you have conflicting converters in your stack — ones using the same encoding, but I'm happy to drop this in case it's not a real concern

Reply:
It's definitely what you want for a payload converter. Because given a collection of payload converters, you may want to choose one based on its encoding (alternatively you can just try each one and have each one return that it can't decode something after it checks encoding metadata key). But the data converter level is higher level and there is just one, it just so happens the default data conversion uses the collection of payload converters which supports fall-through/encoding semantics.

But some of this is admittedly inconsistent between our APIs. But the idea of a highest-level single data converter that has to/from and the default is one that accepts multiple individual encoding-based payload converters is fairly common.